File: app/src/test/java/com/farmerbb/taskbar/ui/UIHostServiceTest.java
package com.farmerbb.taskbar.ui;

import android.content.res.Configuration;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.Shadows;
import org.robolectric.android.controller.ServiceController;
import org.robolectric.shadows.ShadowService;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

@RunWith(RobolectricTestRunner.class)
public class UIHostServiceTest {
    private ServiceController<TestUIHostService> controller;
    private TestUIHostService hostService;
    private TestUIController uiController;

    @Before
    public void setUp() {
        controller = Robolectric.buildService(TestUIHostService.class);
        hostService = controller.create().get();
        uiController = hostService.controller;
    }

    @After
    public void tearDown() {
        uiController.onCreateHost = null;
        uiController.onRecreateHost = null;
        uiController.onDestroyHost = null;
    }

    @Test
    public void testOnCreate() {
        assertEquals(uiController.onCreateHost, hostService);
    }

    @Test
    public void testOnConfigurationChanged() {
        assertNull(uiController.onRecreateHost);
        hostService.onConfigurationChanged(new Configuration());
        assertEquals(hostService, uiController.onRecreateHost);
    }

    @Test
    public void testOnDestroy() {
        assertNull(uiController.onDestroyHost);
        controller.destroy();
        assertEquals(hostService, uiController.onDestroyHost);
    }

    @Test
    public void testTerminate() {
        ShadowService shadowService = Shadows.shadowOf(hostService);
        assertFalse(shadowService.isStoppedBySelf());
        hostService.terminate();
        assertTrue(shadowService.isStoppedBySelf());
    }

}

File: app/src/test/java/com/farmerbb/taskbar/ui/UIHostServiceTest.kt
package com.farmerbb.taskbar.ui

import android.content.res.Configuration
import org.junit.After
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.Robolectric
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows
import org.robolectric.android.controller.ServiceController

@RunWith(RobolectricTestRunner::class)
class UIHostServiceTest {
    private lateinit var controller: ServiceController<TestUIHostService>
    private lateinit var hostService: TestUIHostService
    private lateinit var uiController: TestUIController

    @Before
    fun setUp() {
        controller = Robolectric.buildService(TestUIHostService::class.java)
        hostService = controller.create().get()
        uiController = hostService.controller
    }

    @After
    fun tearDown() {
        uiController.onCreateHost = null
        uiController.onRecreateHost = null
        uiController.onDestroyHost = null
    }

    @Test
    fun testOnCreate() {
        Assert.assertEquals(uiController.onCreateHost, hostService)
    }

    @Test
    fun testOnConfigurationChanged() {
        Assert.assertNull(uiController.onRecreateHost)

        val newConfig = Configuration().apply {
            smallestScreenWidthDp = 123
            screenHeightDp = 456
            screenWidthDp = 789
        }

        val config = uiController.context.resources.configuration.apply {
            updateFrom(newConfig)
        }

        hostService.onConfigurationChanged(config)
        Assert.assertEquals(hostService, uiController.onRecreateHost)
    }

    @Test
    fun testOnDestroy() {
        Assert.assertNull(uiController.onDestroyHost)
        controller.destroy()
        Assert.assertEquals(hostService, uiController.onDestroyHost)
    }

    @Test
    fun testTerminate() {
        val shadowService = Shadows.shadowOf(hostService)
        Assert.assertFalse(shadowService.isStoppedBySelf)
        hostService.terminate()
        Assert.assertTrue(shadowService.isStoppedBySelf)
    }
}
